### PR TITLE
FHO

### PR DESCRIPTION
**Pull request by adel-R**
Good morning Sir,

As part of our optimization project, we have coded a metaheuristic optimization algorithm called Fire Hawk Optimizer that we would like to submit on your github.

To accomplish this, we have basically followed the steps of the original article from Springer : https://link.springer.com/article/10.1007/s10462-022-10173-w#Sec2

So far, our algorithm seems to function very well on functions we have tested:
Sphere                   2-D & 25-D
Exponential          2-D & 25-D          
Ackley                    2-D & 25-D
Becker-Lago          2-D & 25-D
Bird                        2-D only    
Example below for : Ackley function in 25-D (plots in 2-D are for  illustrative purposes)

Obtained solution : 
f(x) = 0.02825112986538514
xmin = array([-7.24768449e-03, 
                       -1.75967133e-03, 
                       -1.59870557e-03, 
                         4.61922463e-03,         
                        -5.69497292e-05, 
                        -1.26671059e-04,  
                        7.63080036e-03, 
                         3.09565889e-03,          
                        4.03418283e-04,  
                        6.30693074e-03, 
                        -1.92647450e-03,  
                        5.90686009e-03,         -
                        1.24540399e-02,  
                        9.04388929e-04, 
                        -1.73015287e-03,  
                        1.61377715e-02,         
                        -3.06922319e-04,  
                        2.23402266e-03,  
                        6.40657103e-03,  
                        1.15464648e-02,        
                         -1.37048453e-02,  
                        3.81699619e-03, 
                        -3.35189100e-03, 
                        -2.50757084e-03,         
                        -3.44429387e-03])

Best regards,

Aiza, Xingshan, Natacha, Vanshika, Hamza and Adel

Closes # 

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

File: models/multiple_solution/swarm_based/FHO.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                    PRq_new1= np.maximum(PRq_new1,minD)
+                    PRq_new1 = np.minimum(PRq_new1,maxD)
+                    PopTot.append(list(PRq_new1))
+                    
+                    #Movement of the preys outside of territory
+                    a,b      =np.random.uniform(0,1,size=2)
+                    FHAlter  =Fhawks[np.random.randint(num_Hawks)] 
+                    PRq_new2 =PR[q].copy()+((a*FHAlter-b*SP));
+                    PRq_new2 = np.maximum(PRq_new2,minD)
+                    #The following line for PRq_new2 differs from original algorithm in matlab code (max instead of min):
```
Comment (thieu1995) [inline models/multiple_solution/swarm_based/FHO.py:133]: Just implement the equations as in the original paper. Many people they are faking the results of paper in the Matlab code. If you change the equations by your opinion, it is not their algorithm anymore. Rename this class as OriginalFHO and implement it as exactly same as in the original paper. You can create another class in the file name DevFHO (DevelopedFHO), based on your opinion. Compare two versions and see the results.

**Review by thieu1995 — CHANGES_REQUESTED** [review]

**Comment by adel-R** [comment]
Hi,

Following your last comments, all the variables have been renamed:
N -> pop_size
Xcandidates -> Pop
d -> n_dims

And all the camel case have been changed to _ notations.

In addition, following your comment, two classes exist in the FHO file :
-> Developed_FHO (including our modification)
-> Original_FHO (which sticks to the matlab code of the author).

Please note that the modified line is not explicitly covered in the article and may be a typo from the authors....

In all our tests (see notebook), the algorithm converged to one of the global solutions...

File: models/multiple_solution/swarm_based/FHO.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    self.cost_function: fucntion to optimize
+    self.candidates : List of Candidates
+    self.d: dimension of the problem
+    self.Fhawks : list of fire hawks
+    self.preys : list of preys
+    self.territories : territory of each hawk
+    self.N : Number of solution candidates (int)
+    self.minD : numpy array of dimension (d,) containing the MIN value for each variable
+    self.maxD : numpy array of dimension (d,) containing the MAX value for each variable
+    self.N : Number of solution candidates (int)
```
[the remark this answers is not in the document]
Comment (adel-R) [reply models/multiple_solution/swarm_based/FHO.py]: Done

File: models/multiple_solution/swarm_based/FHO.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    self.best_costs: list of best costs
+    """
+    
+    def __init__(self, minD , maxD, N, function , max_generations=200):
+        self.max_generations = max_generations
+        self.minD            = minD
+        self.maxD            = maxD
+        self.cost_function   = function
+        self.N               = N
+        self.d               = len(minD) #Number of decision variables
```
[the remark this answers is not in the document]
Comment (adel-R) [reply models/multiple_solution/swarm_based/FHO.py]: Done

File: models/multiple_solution/swarm_based/FHO.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        self.Xcandidates     = np.random.uniform(minD,maxD,(N,self.d)) #Initial Solution candidates (Nxd)
```
[the remark this answers is not in the document]
Comment (adel-R) [reply models/multiple_solution/swarm_based/FHO.py]: Done

File: models/multiple_solution/swarm_based/FHO.py (hunk 1 above)
Comment (adel-R) [reply models/multiple_solution/swarm_based/FHO.py:147]: Following your comment, two classes have been created in the FHO.py file :
-> Developed_FHO (including our modification)
-> Original_FHO (which sticks to the matlab code of the author).

Please note that the modified line is not explicitly covered in the article and may be a typo from the authors....

In all our tests (see notebook "FHO_VF.ipynb"), our algorithm converged to one of the global solutions... While the 'original' does not converge in 25-D for the Sphere function...

**Comment by adel-R** [comment]
run_FHO.py has been committed  for both classes!

**Pull request merged**